### ISSUE TITLE
Memory Leaks in WebClient Get Request with Body

### ISSUE THREAD

**Issue by codeomnitrix** [spring-projects/spring-framework#24082]
Hi,

When I try to make a get request with body using webclient, it creates memory leaks in direct memory. And 10-12 hours of running the service we see following exception - 
 
```
Direct buffer memory
	at java.base/java.nio.Bits.reserveMemory(Unknown Source)
	at java.base/java.nio.DirectByteBuffer.<init>(Unknown Source)
	at java.base/java.nio.ByteBuffer.allocateDirect(Unknown Source)
	at io.netty.buffer.PoolArena$DirectArena.allocateDirect(PoolArena.java:768)
	at io.netty.buffer.PoolArena$DirectArena.newChunk(PoolArena.java:744)
	at io.netty.buffer.PoolArena.allocateNormal(PoolArena.java:245)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:215)
	at io.netty.buffer.PoolArena.allocate(PoolArena.java:147)
	at io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:327)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:187)
	at io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:173)
	at io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:107)
	at org.springframework.core.io.buffer.NettyDataBufferFactory.allocateBuffer(NettyDataBufferFactory.java:65)
	at org.springframework.core.io.buffer.NettyDataBufferFactory.allocateBuffer(NettyDataBufferFactory.java:39)
	at org.springframework.http.codec.json.AbstractJackson2Encoder.encodeValue(AbstractJackson2Encoder.java:168)
	at org.springframework.http.codec.json.AbstractJackson2Encoder.lambda$encode$0(AbstractJackson2Encoder.java:122)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:107)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.onNext(FluxHide.java:127)
	at reactor.core.publisher.Operators$ScalarSubscription.request(Operators.java:2070)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.request(FluxHide.java:142)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:162)
	at org.springframework.cloud.sleuth.instrument.reactor.ScopePassingSpanSubscriber.request(ScopePassingSpanSubscriber.java:79)
	at reactor.core.publisher.FluxHide$SuppressFuseableSubscriber.request(FluxHide.java:142)
	at reactor.core.publisher.MonoSingle$SingleSubscriber.request(MonoSingle.java:94)
	at org.springframework.cloud.sleuth.instrument.reactor.ScopePassingSpanSubscriber.request(ScopePassingSpanSubscriber.java:79)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.set(Operators.java:1878)
	at reactor.core.publisher.Operators$MultiSubscriptionSubscriber.onSubscribe(Operators.java:1752)
	at org.springframework.cloud.sleuth.instrument.reactor.ScopePassingSpanSubscriber.onSubscribe(ScopePassingSpanSubscriber.java:71)
```

Code for the get with body is as follows- 

```
public Mono<List<ReviewDetailsV1>> getReviewsByPostData(SearchCriteria searchCriteria, Mono<PropertyReviewsRequestBody> propertyReviewsRequestBody) {
        CircuitBreaker reviewCircuitBreaker = circuitBreakerRegistry.circuitBreaker("property-reviews-webclient");
        return this.webClient
                .method(HttpMethod.GET)
                .uri(url -> UrlProviderUtil.getUrlForReviews(url, searchCriteria))
                .contentType(MediaType.APPLICATION_JSON_UTF8)
                .header(HttpHeadersEnum.REQUEST_ID.getHeader(), MDC.get(MdcKeysEnum.REQUEST_ID.getMdcKey()))
                .header(HttpHeadersEnum.SESSION_ID.getHeader(), MDC.get(MdcKeysEnum.SESSION_ID.getMdcKey()))
                .header(HttpHeadersEnum.DEBUG_HEADER.getHeader(), DEBUG.equalsIgnoreCase(MDC.get(MdcKeysEnum.DEBUG_REQUEST.getMdcKey())) ? DEBUG : "")
                .body(propertyReviewsRequestBody, PropertyReviewsRequestBody.class)
                .retrieve()
                .bodyToFlux(ReviewDetailsV1.class)
                .transform(CircuitBreakerOperator.of(reviewCircuitBreaker))
                .collectList()
                .doOnSuccess(res -> longOnReviewsDataSuccess(searchCriteria, res))
                .onErrorResume(ex -> reviewsFallback(searchCriteria, propertyReviewsRequestBody, ex));
    }
```

However if I comment the line `.body(propertyReviewsRequestBody, PropertyReviewsRequestBody.class)` it works fine. 

Attached graphs for direct memory allocated.
<img width="1646" alt="Screen Shot 2019-11-26 at 4 53 06 PM" src="https://user-images.githubusercontent.com/4947044/69625917-9675be80-106d-11ea-8f36-99c8c1d8266a.png">

Is it a known issue?

Spring Boot Dependency used - 
```
<dependency>
        <groupId>org.springframework.boot</groupId>
        <artifactId>spring-boot-starter-webflux</artifactId>
        <version>2.1.6.RELEASE</version>
      </dependency>
```

**Comment by rstoyanchev** [comment]
The first thing to try is to check with Boot 2.1.10 and 2.2.1. If that doesn't help the best option would be having some way to reproduce the issue.

**Comment by rameshgkaranth** [comment]
@rstoyanchev Is GET with Body supported use case in WebClient? 
We found lot of instances of body object in DirectByteBuffer, should the body be drained from byte buffers? If yes, can you tell us how we can do the same?

**Comment by rstoyanchev** [comment]
> Is GET with Body supported use case in WebClient?

I don't think this has anything to do with the HTTP method, and you probably would have seen other issues if it wasn't supported. You can try and switch to a different HTTP method, if that's feasible, just to prove it. I suspect that won't make a difference.

More likely the circuit break is causing the request to be cancelled from time to time and there might something in the encoder layer that's not releasing the buffers. What is the request body getting serialized to, i.e. what is the content type, and how is the encoding layer configured?

**Comment by rameshgkaranth** [comment]
ContentType is application/json. `PropertyReviewsRequestBody` is an object that has a Map<String, Long>. We do not have additional encoding layer configured apart from what you see in the snippet above.

We do have occasional circuit breaker errors due to timeouts, the method used in `onErrorResume ` calls `reviewsFallback`, it is used as a fallback to log error and return empty list.

```private Mono<List<ReviewDetailsV1>> reviewsFallback(SearchCriteria searchCriteria, Mono<PropertyReviewsRequestBody> propertyReviewsRequestBody, Throwable e) {
    LOGGER.error("...");
    return Mono.just(new ArrayList<>());
}```

**Comment by rstoyanchev** [comment]
We did  have some issues fixed in the 5.1.x cycle, so please try the latest 2.1.10 (or 2.1.0).

**Comment by codeomnitrix** [comment]
@rstoyanchev upgrading to 2.1.10 did the trick. I think there was a bug in 2.1.6 version with direct memory leak. Closing the bug.
